### DISCUSSION TITLE
Help to get this working

### DISCUSSION THREAD

**Discussion by MeestorX** [exposedev/expose#229] [Q&A]
It sounds like at least some of you have been able to get this working?

I would like to create my own expose server to tunnel a TCP Port on a local computer just like nGrok can do.
I created a DO Docker droplet.
Installed expose
Not sure what to do next?

What app do I download to the client? With nGrok there's a Windows/Mac/Linux app for the local computer that's going to share it's port.

What command(s) do I run on my server?

I've gone through the documentation many times and I'm missing a few key steps...

Thanks for taking pity on me, anyone kind enough to help! :-)

**Comment by cpxPratik** [comment]
You can install and use Expose client(on your local machine) as follows:

## Installing Expose client

-   Install Expose as global composer dependency

```bash
composer global require beyondcode/expose
```

-   If composer's vendor bin direcotry is not added on PATH environment variable, add following line at the end of .bashrc or .zshrc file located on your home directory.

On Linux system,

```bash
PATH=$PATH:~/.config/composer/vendor/bin
```

On macOS,

```bash
PATH=$PATH:~/.composer/vendor/bin
```

-   Reload .bashrc or .zshrc by running `source ~/.bashrc` or `source ~/.zshrc`


-   Run `expose publish` to publish Expose configuration file at ~/.expose/config.php. Refer <https://beyondco.de/docs/expose/client/configuration> for more details.


-   Update host value to `your_expose_domain` on file `~/.expose/config.php`
-   Ask Expose admin for authentication token to connect to `your_expose_domain` server.
-   Run `expose token <your_auth_token>` . It updates `~/.expose/config.php` with token value.

## Using Expose client

-   Run your app on your local machine then share it with `expose share http://127.0.0.1:<local_app_port> --subdomain=<sub_domain>`

```bash
expose share http://127.0.0.1:<local_port> --subdomain=rest-skeleton
```

-   Refer <https://beyondco.de/docs/expose/client/sharing> for more details


-   To share local site with custom basic authentication credentials, add parameter with `--auth="<basic_auth_username>:<basic_auth_password`

```bash
expose share http://127.0.0.1:<local_port> --subdomain=rest-skeleton --auth="admin:test123456"
```

**Comment by MeestorX** [comment]
@cpxPratik thank you so much!
Gave up on the docker and used composer. The PATH info was missing from the instructions, particularly for MacOS.

now, how does one share a TCP port?
On Ngrok it’s: `ngrok tcp <ip>:<port>`

I tried:
`expose share-port <ip>:<port> --subdomain=andy` 

It does not work, as per https://github.com/beyondcode/expose/pull/123#issuecomment-706661112

**Comment by MeestorX** [comment]
Anyone?

**Comment by mpociot** [comment]
TCP port sharing is only available in the master branch and will land in version 2.x. You can already set up your own server to use this though

**Comment by MeestorX** [comment]
Go on...

That is my goal to use my own server. Do you have instructions on how to do TCP port sharing on my own server?

**Comment by ghostzero** [comment]
There is a documentation available that explain how to host your own server. You may want to check this out:
https://beyondco.de/docs/expose/server/starting-the-server

**Comment by MeestorX** [comment]
Right, I've read that. I'm talking about instructions on how to share TCP ports specifically.

**Comment by MeestorX** [comment]
Anyone?

**Comment by MeestorX** [comment]
Friendly bump in case anyone’s figured out how to share TCP ports yet?

**Comment by ghost** [comment]
Yeah this is a bit frustrating.

You guys are so innovative and productive, but you need to hire someone to write documentation.

I created a Forge server, added my domain, installed SSL.

I downloaded the package and then created a repo for it, then imported it as an app into my domain on Forge.

I adjusted the config to represent the correct domain and port 443, I added your configuration for Nginx to map / to 127.0.0.1:8080. 

I opened port 8080 just in case it was needed. I created a daemon to run `php expose serve mydomain.com` 

I visit the site and a see an SSL encrypted homepage. So far so good.

Now, locally I installed the package globally:

```bash
composer require -g beyondcode/expose
```

I publish the config and edit it accordingly, adding the domain of my server and port 443

And then .... 

when I run `expose share mysite.test --subdomain=example` I get nothing ... no feedback, no error nothing.

It just hangs forever.

So .... I think ... maybe I need to start a local server for it to work? So I run:

```bash
expose serve mydomain.com
```
Expose reports that it's running on port 8080 ... which is NOT what's in the config, but either way.

Still, trying to share a site produces nothing at all, just a hanging process.

Meanwhile ... I have a server burning that does nothing.

What's the hold up on these issues?

**Comment by warren-bank** [comment]
fwiw, [here](https://github.com/warren-bank/HLS-Proxy/tree/master/.recipes/02.%20NAT%20tunnel/01.%20expose) is a fully working example that I used to test using _expose_ to tunnel an _HLS Proxy_ server.

I didn't see any need to use _Docker_ or _composer_. The _expose_ repo includes a [_PHAR_](https://github.com/beyondcode/expose/raw/1.5.1/builds/expose) that only needs a _PHP_ binary on _PATH_ to run.

* scripts are written for Windows _cmd_ (not _bash_, sorry)
* test scenario:
  - update the system's [_hosts_](https://github.com/warren-bank/HLS-Proxy/blob/master/.recipes/02.%20NAT%20tunnel/01.%20expose/data/hosts) file
  - run [_HLS Proxy_ server](https://github.com/warren-bank/HLS-Proxy/blob/master/.recipes/02.%20NAT%20tunnel/01.%20expose/1.%20run%20hls-proxy%20server.bat) on _localhost:8080_
  - run [_expose_ server](https://github.com/warren-bank/HLS-Proxy/blob/master/.recipes/02.%20NAT%20tunnel/01.%20expose/2.%20run%20expose%20server.bat) on _my-domain:9090_
    * in this scenario, the updated system _hosts_ file forces DNS to resolve _my-domain_ to _localhost_
    * in a real-world scenario, a registered domain would be used that will be resolved by a public DNS server
  - run [_expose_ client](https://github.com/warren-bank/HLS-Proxy/blob/master/.recipes/02.%20NAT%20tunnel/01.%20expose/4.%20run%20expose%20client.bat) on same machine as _HLS Proxy_ server
    * endpoint of tunnel is on same machine as _expose_ server
    * it is accessed at the static subdomain _hls-proxy.my-domain:9090_
